lib/pure/osproc.nim

```diff
@@ -434,6 +434,87 @@ when not defined(useNimRtl):
 template streamAccess(p) =
   assert poParentStreams notin p.options, "API usage error: stream access not allowed when you use poParentStreams"
 
+# Wrap pipe for reading with PipeOutStream so that you can use peek* and generate runtime error
+# when setPosition/getPosition is called or write operation is performed.
+type
+  PipeOutStream[T] = ref object of T
+    buffer: Deque[char]
+    baseReadLineImpl: typeof(StreamObj.readLineImpl)
+    baseReadDataImpl: typeof(StreamObj.readDataImpl)
+
+proc posReadLine[T](s: Stream, line: var TaintedString): bool =
+  var s = PipeOutStream[T](s)
+  assert s.baseReadLineImpl != nil
+
+  let n = s.buffer.len
+  line.string.setLen(0)
+  for i in 0..<n:
+    var c = s.buffer.popFirst
+    if c == '\c':
+      c = readChar(s)
```

Comment:
I copied code from `readLine` proc in `streams` module so that `posReadLine` take new line from the buffer in the same way as `readLine` in `streams` module.
On windows and dos, a line of text is delimited ``CRLF``.
The newline character(s) are not part of the returned string.